### PR TITLE
Speedup CI testing

### PR DESCRIPTION
**Pull request by oke-aditya**
## What does this PR do?

Speeds up CI Testing.

Some optmizations followed

- [x] Use `@torch.no_grad()` to avoid gradients.
- [x] Avoid Downloading Pretrained Weights, These are provided by torchvision we can rely on them and save time.
- [x] Use `fast_dev_run`, disable logging, checkpoint callbacks.

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning-bolts/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? [not needed for typos/docs]
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning-bolts/blob/master/CHANGELOG.md)?

## PR review
 - [x] Is this pull request ready for review? (if not, please submit in draft mode)

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/models/self_supervised/test_resnets.py [hunk 1]
```diff
@@ -16,6 +16,7 @@
 )
 
 
+@torch.no_grad()
```
Comment (oke-aditya) [inline tests/models/self_supervised/test_resnets.py:19]: Since for all the resnet models we don't compute backward. We shouldn't keep track of gradients, :smile:

File: tests/models/test_gans.py [hunk 2]
```diff
@@ -18,7 +18,9 @@ def test_gan(tmpdir, datadir, dm_cls):
 
     dm = dm_cls(data_dir=datadir, num_workers=0)
     model = GAN(*dm.size())
-    trainer = pl.Trainer(fast_dev_run=True, default_root_dir=tmpdir)
+    trainer = pl.Trainer(
+        fast_dev_run=True, logger=False, checkpoint_callback=False, max_epochs=1, default_root_dir=tmpdir
```
Comment (oke-aditya) [inline tests/models/test_gans.py:22]: I think we can standardize to use something like this.

File: tests/models/test_vision.py [hunk 3]
```diff
@@ -7,32 +7,36 @@
...
-        limit_val_batches=2,
-        limit_test_batches=2,
-        max_epochs=1,
+        fast_dev_run=True, logger=False, checkpoint_callback=False, max_epochs=1, default_root_dir=tmpdir
     )
+
     trainer.fit(model, datamodule=dm)
 
 
+@torch.no_grad()
```
Comment (oke-aditya) [inline tests/models/test_vision.py:39]: No grad needed here too. (Might save a lot GPT-2 with no grad :thinking: )

File: tests/models/self_supervised/test_resnets.py [hunk 4]
```diff
@@ -35,12 +36,14 @@ def test_cpc_resnet():
         wide_resnet101_2,
     ]
 )
+@torch.no_grad()
```
Comment (oke-aditya) [inline tests/models/self_supervised/test_resnets.py:39]: This can save a bunch of time, since it avoids totally a lot of models

**Comment by oke-aditya** [comment]
I think before merging, we can check if we are really getting speedup. 
I think it's very little around 2 mins or so in CI Full testing, latest. Disabling grads is giving slight benefit.

File: tests/models/test_gans.py (hunk 2 above)
Comment (Borda) [reply tests/models/test_gans.py:22]: is it needed when you already have `fast_dev_run=True`?

File: tests/models/test_vision.py [hunk 5]
```diff
@@ -6,33 +6,32 @@
 from pl_bolts.datasets import DummyDataset
 from pl_bolts.models import GPT2, ImageGPT, SemSegment, UNet
 
+pl.seed_everything(42)
```
Comment (Borda) [inline tests/models/test_vision.py:9]: this shall be inside functions

File: tests/models/test_vision.py [hunk 6]
```diff
@@ -50,26 +49,22 @@ def test_gpt2():
...
 
-def test_semantic_segmentation():
-
-    class DummyDataModule(pl.LightningDataModule):
-
-        def train_dataloader(self):
-            train_ds = DummyDataset((3, 35, 120), (35, 120), num_samples=100)
-            return DataLoader(train_ds, batch_size=1)
-
+def test_semantic_segmentation(tmpdir, ):
```
Comment (Borda) [inline tests/models/test_vision.py:60]:
```suggestion
def test_semantic_segmentation(tmpdir):
```

File: tests/models/test_vision.py [hunk 7]
```diff
@@ -6,33 +6,32 @@
 from pl_bolts.datasets import DummyDataset
 from pl_bolts.models import GPT2, ImageGPT, SemSegment, UNet
 
+pl.seed_everything(42)
 
-def test_igpt(datadir):
-    pl.seed_everything(0)
```
Comment (akihironitta) [inline tests/models/test_vision.py:11]: As long as we keep this seed, `test_igpt` passes with the current loss value `1.7`, right?

Comment (oke-aditya) [reply tests/models/test_vision.py:11]: Yes, but that's unreliable testing, I mean the loss should rather satisfy some value irrespective of seed.

File: tests/models/test_gans.py (hunk 2 above)
Comment (oke-aditya) [reply tests/models/test_gans.py:22]: Well yes, This avoid some calls to logger, prevents checkpoints being stored.

Checkpoints can occupy unnecassary disk space and possible transfer from RAM to Memory causing `Memory Error`

I'm not sure if this will exactly solve the problem #526 But keeping this enabled will probably avoid these issues.

File: tests/models/test_vision.py (hunk 5 above)
Comment (oke-aditya) [reply tests/models/test_vision.py:9]: Same comment as above. This is defensive programming, to avoid unnessary errors later :smiley:

Comment (oke-aditya) [reply tests/models/test_vision.py:9]: I tihnk globally seeding will avoid duplication of code and possible calls?
We can be determinstic with a seed say 0 or 42 (scikit-learn :smiley: number)

**Comment by oke-aditya** [comment]
I will change the issue name ready to merge, when optimization is good to go :smile: 

I will post a detailed comparison of optimization here, I will compare master runs with this.

File: tests/models/test_vision.py (hunk 7 above)
Comment (akihironitta) [reply tests/models/test_vision.py:11]: Yeah, you're right... but I'm still not sure if specifying a certain value of the loss is reasonable because the value is based on the implementation which we want to test, right...? But for now, I guess we could just change this value.

**Comment by oke-aditya** [comment]
Hey @Borda  and @akihironitta , I have done optimizations and this is good to go.
I see no codecoverage dips, nor there is any CI failure :smile:  :rocket:

File: tests/models/test_gans.py (hunk 2 above)
Comment (akihironitta) [reply tests/models/test_gans.py:22]: Do we need `max_epochs=1` here? `fast_dev_run=True` ends right after the first batch in the first epoch, right?

**Review by akihironitta — APPROVED** [review]
Other than @Borda's comment:
> is it needed when you already have `fast_dev_run=True`?

It looks good to me.

**Comment by oke-aditya** [comment]
Let me remove `max_epochs` with `fast_dev_run`

**Comment by oke-aditya** [comment]
@Borda  can you review :smiley:

File: tests/models/self_supervised/test_models.py [hunk 8]
```diff
@@ -40,7 +40,7 @@ def test_byol(tmpdir, datadir):
     datamodule.val_transforms = CPCEvalTransformsCIFAR10()
 
     model = BYOL(data_dir=datadir, num_classes=datamodule)
-    trainer = pl.Trainer(fast_dev_run=True, max_epochs=1, default_root_dir=tmpdir, max_steps=2)
+    trainer = pl.Trainer(fast_dev_run=True, default_root_dir=tmpdir, max_steps=2)
```
Comment (akihironitta) [inline tests/models/self_supervised/test_models.py:43]:
```suggestion
    trainer = pl.Trainer(fast_dev_run=True, default_root_dir=tmpdir)
```

File: tests/models/self_supervised/test_models.py [hunk 9]
```diff
@@ -119,7 +119,7 @@ def test_swav(tmpdir, datadir):
         dataset='cifar10'
     )
 
-    trainer = pl.Trainer(gpus=0, fast_dev_run=True, max_epochs=1, default_root_dir=tmpdir, max_steps=3)
+    trainer = pl.Trainer(gpus=0, fast_dev_run=True, default_root_dir=tmpdir, max_steps=3)
```
Comment (akihironitta) [inline tests/models/self_supervised/test_models.py:122]:
```suggestion
    trainer = pl.Trainer(gpus=0, fast_dev_run=True, default_root_dir=tmpdir)
```

File: tests/models/test_classic_ml.py [hunk 10]
```diff
@@ -20,7 +20,7 @@ def test_linear_regression_model(tmpdir):
     loader = DataLoader(SklearnDataset(X, y), batch_size=2)
 
     model = LinearRegression(input_dim=2, learning_rate=0.5)
-    trainer = pl.Trainer(max_epochs=300, default_root_dir=tmpdir, progress_bar_refresh_rate=0)
+    trainer = pl.Trainer(max_epochs=300, default_root_dir=tmpdir, progress_bar_refresh_rate=0, logger=False)
```
Comment (akihironitta) [inline tests/models/test_classic_ml.py:23]: Let's turn off saving checkpoints here since we're not using `fast_dev_run=True`. [the docs](https://pytorch-lightning.readthedocs.io/en/stable/debugging.html#fast-dev-run)

```suggestion
    trainer = pl.Trainer(max_epochs=300, default_root_dir=tmpdir, progress_bar_refresh_rate=0, logger=False, checkpoint_callback=False)
```

File: tests/models/test_classic_ml.py [hunk 11]
```diff
@@ -46,7 +46,12 @@ def test_logistic_regression_model(tmpdir, datadir):
     model.val_dataloader = dm.val_dataloader
     model.test_dataloader = dm.test_dataloader
 
-    trainer = pl.Trainer(max_epochs=3, default_root_dir=tmpdir, progress_bar_refresh_rate=0)
+    trainer = pl.Trainer(
+        max_epochs=3,
+        default_root_dir=tmpdir,
+        progress_bar_refresh_rate=0,
+        logger=False,
+    )
```
Comment (akihironitta) [inline tests/models/test_classic_ml.py:54]: Same as above.
```suggestion
        logger=False,
        checkpoint_callback=False,
    )
```

File: tests/models/test_detection.py [hunk 12]
```diff
@@ -10,27 +10,28 @@ def _collate_fn(batch):
...
 
 def test_fasterrcnn_train(tmpdir):
-    model = FasterRCNN()
+    model = FasterRCNN(pretrained=False, pretrained_backbone=False)
 
     train_dl = DataLoader(DummyDetectionDataset(), collate_fn=_collate_fn)
     valid_dl = DataLoader(DummyDetectionDataset(), collate_fn=_collate_fn)
 
-    trainer = pl.Trainer(fast_dev_run=True, default_root_dir=tmpdir)
+    trainer = pl.Trainer(fast_dev_run=True, logger=False, checkpoint_callback=False, default_root_dir=tmpdir)
```
Comment (akihironitta) [inline tests/models/test_detection.py:27]: I like your idea of being defensive, but lightning ensures to disable these options with `fast_dev_run=True` as stated in [the docs](https://pytorch-lightning.readthedocs.io/en/stable/debugging.html#fast-dev-run) so how about keeping this line as is? 
```suggestion
    trainer = pl.Trainer(fast_dev_run=True, default_root_dir=tmpdir)
```

File: tests/models/test_detection.py [hunk 13]
```diff
@@ -10,27 +10,28 @@ def _collate_fn(batch):
...
 
 
 def test_fasterrcnn_bbone_train(tmpdir):
-    model = FasterRCNN(backbone="resnet18", fpn=True, pretrained_backbone=True)
+    model = FasterRCNN(backbone="resnet18", fpn=True, pretrained_backbone=False, pretrained=False)
     train_dl = DataLoader(DummyDetectionDataset(), collate_fn=_collate_fn)
     valid_dl = DataLoader(DummyDetectionDataset(), collate_fn=_collate_fn)
 
-    trainer = pl.Trainer(fast_dev_run=True, default_root_dir=tmpdir)
+    trainer = pl.Trainer(fast_dev_run=True, logger=False, checkpoint_callback=False, default_root_dir=tmpdir)
```
Comment (akihironitta) [inline tests/models/test_detection.py:36]: Same as above.
```suggestion
    trainer = pl.Trainer(fast_dev_run=True, default_root_dir=tmpdir)
```

File: tests/models/test_gans.py [hunk 14]
```diff
@@ -18,7 +18,7 @@ def test_gan(tmpdir, datadir, dm_cls):
 
     dm = dm_cls(data_dir=datadir, num_workers=0)
     model = GAN(*dm.size())
-    trainer = pl.Trainer(fast_dev_run=True, default_root_dir=tmpdir)
+    trainer = pl.Trainer(fast_dev_run=True, logger=False, checkpoint_callback=False, default_root_dir=tmpdir)
```
Comment (akihironitta) [inline tests/models/test_gans.py:21]:
```suggestion
    trainer = pl.Trainer(fast_dev_run=True, default_root_dir=tmpdir)
```

File: tests/models/test_gans.py [hunk 15]
```diff
@@ -34,6 +34,6 @@ def test_dcgan(tmpdir, datadir, dm_cls):
     dm = dm_cls(data_dir=datadir, train_transforms=transforms, val_transforms=transforms, test_transforms=transforms)
 
     model = DCGAN(image_channels=dm.dims[0])
-    trainer = pl.Trainer(fast_dev_run=True, default_root_dir=tmpdir)
+    trainer = pl.Trainer(fast_dev_run=True, logger=False, checkpoint_callback=False, default_root_dir=tmpdir)
```
Comment (akihironitta) [inline tests/models/test_gans.py:37]:
```suggestion
    trainer = pl.Trainer(fast_dev_run=True, default_root_dir=tmpdir)
```

File: tests/models/test_vision.py [hunk 16]
```diff
@@ -50,26 +61,20 @@ def test_gpt2():
...
-            train_ds = DummyDataset((3, 35, 120), (35, 120), num_samples=100)
-            return DataLoader(train_ds, batch_size=1)
-
+def test_semantic_segmentation(tmpdir):
     dm = DummyDataModule()
 
     model = SemSegment(num_classes=19)
 
-    trainer = pl.Trainer(fast_dev_run=True, max_epochs=1)
+    trainer = trainer = pl.Trainer(fast_dev_run=True, logger=False, checkpoint_callback=False, default_root_dir=tmpdir)
```
Comment (akihironitta) [inline tests/models/test_vision.py:77]:
```suggestion
    trainer = trainer = pl.Trainer(fast_dev_run=True, default_root_dir=tmpdir)
```

**Review by akihironitta — COMMENTED** [review]
Hey @oke-aditya, I think this is a great improvement in the tests! The comments below are all from me, so once these are agreed upon between us, this PR is good to go :)

**Comment by oke-aditya** [comment]
Hey @akihironitta @Borda  I have done the changes :smiley: 

Sorry for the delay, last few weeks were tedious

**Review by akihironitta — APPROVED** [review]

**Review by Borda — APPROVED** [review]
